### PR TITLE
Feature/store active games in database

### LINKED ISSUE #86: Store active games in database

### PR DESCRIPTION
**Pull request by mattiasnilsen**
I have moved all information stored about a game into the database.

All functionality should work as expected however the tests does not always pass, they fail with an error like: `*** Failed! Exception: 'ConnectionError {errFunction = "query", errNumber = 1205, errMessage = "Lock wait timeout exceeded; try restarting transaction"}'`

Because of this I have not created any new tests, more investigation is needed into correctly test code where the database is involved.

fixes #86

### CODE REVIEW

**Comment by Kalior** [comment]
I would say tests need to be written, since you changed how some of the core functionality of the lobby works.

File: app/StandaloneApp.hs [hunk 1]
```diff
@@ -14,20 +14,25 @@ import GameAPI
...
   chatList <- liftServerIO $ CC.newMVar []
 
-  let serverState = (playersList, gamesList, chatList)
+  let serverState = (playersList, chatList)
+  liftServerIO $ migrateDatabase
+  liftServerIO $ clearOnlinePlayers
 
   onSessionEnd $ disconnect(serverState)
-  api <- newLobbyAPI (playersList, gamesList, chatList)
+  api <- newLobbyAPI (playersList, chatList)
```
Comment (Kalior) [inline app/StandaloneApp.hs]: why not use serverState?

File: src/Hastings/Database/Fields.hs [hunk 2]
```diff
@@ -13,16 +13,29 @@ module Hastings.Database.Fields where
...
 share [mkPersist sqlSettings, mkMigrate "migrateAll"] [persistLowerCase|
 Game
-    name               String
-    maxAmountOfPlayers Int
+    uuid                String
+    name                String
+    maxAmountOfPlayers  Int
+    owner               SessionID
+    password            ByteString
+    UniqueName          name
```
Comment (Kalior) [inline src/Hastings/Database/Fields.hs]: Why are names unique?

File: src/Hastings/Database/Game.hs [hunk 3]
```diff
@@ -0,0 +1,129 @@
...
+
+import Haste.App (SessionID)
+import Data.Maybe (listToMaybe)
+import Control.Monad (liftM)
+import Data.ByteString.Char8 (ByteString)
+
+-- |Retrieve a game from the database.
+retrieveGameByName :: String -- ^The name of the game to retrieve.
+                   -> IO (Maybe (Esql.Entity Game))
+retrieveGameByName name = runDB $ Esql.getBy $ UniqueName name
```
Comment (Kalior) [inline src/Hastings/Database/Game.hs]: If Names are not unique, remove this function.

File: src/Hastings/Database/Fields.hs [hunk 4]
```diff
@@ -13,16 +13,29 @@ module Hastings.Database.Fields where
...
+    UniqueUUID          uuid
+    deriving Show
+PlayerInGame
+    game          GameId
+    player        SessionID
```
Comment (Kalior) [inline src/Hastings/Database/Fields.hs:29]: add ID after game and player?

File: src/Hastings/Database/Game.hs [hunk 5]
```diff
@@ -0,0 +1,129 @@
...
+             -> SessionID     -- ^The sessionID of the new owner.
+             -> IO ()
+setGameOwner gameKey sid = runDB $
+  Esql.update $ \games -> do
+    Esql.set games [GameOwner Esql.=. Esql.val sid]
+    Esql.where_ (games Esql.^. GameId Esql.==. Esql.val gameKey)
+
+-- |Add a player to a game
+addPlayerToGame :: SessionID -- ^The sessionID of the player.
+                -> Esql.Key Game -- ^The key of the game.
```
Comment (Kalior) [inline src/Hastings/Database/Game.hs:81]: Align comments please :computer:

File: src/Hastings/Database/Game.hs [hunk 6]
```diff
@@ -0,0 +1,129 @@
...
+-- |Remove a player from a game
+removePlayerFromGame :: SessionID -- ^The sessionID od the player.
+                     -> Esql.Key Game -- ^The key of the game.
+                     -> IO ()
+removePlayerFromGame sessionID gameKey = runDB $
+  Esql.delete $ Esql.from $ \playersInGame ->
+    Esql.where_ (playersInGame Esql.^. PlayerInGameGame Esql.==. Esql.val gameKey
+        Esql.&&. playersInGame Esql.^. PlayerInGamePlayer Esql.==. Esql.val sessionID)
+
+-- |Retrieves a players that are currently in a game.
```
Comment (Kalior) [inline src/Hastings/Database/Game.hs]: a players?

File: src/Hastings/Database/Game.hs [hunk 7]
```diff
@@ -0,0 +1,129 @@
...
+retrievePlayersInGame gameKey = runDB $
+      Esql.select $ Esql.from $ \players -> do
+        (playerInGame, onlinePlayers) <- onlinePlayerInGame
+        Esql.where_ (onlinePlayers Esql.^. OnlinePlayerPlayer Esql.==. players Esql.^. PlayerId
+            Esql.&&. playerInGame Esql.^. PlayerInGameGame Esql.==. Esql.val gameKey)
+        return players
+  where
+    onlinePlayerInGame = Esql.from $ \(playerInGame `Esql.InnerJoin` onlinePlayers) -> do
+        Esql.on (playerInGame Esql.^. PlayerInGamePlayer Esql.==. onlinePlayers Esql.^. OnlinePlayerSessionID)
+        return (playerInGame, onlinePlayers)
```
Comment (Kalior) [inline src/Hastings/Database/Game.hs:108]: Please comment this function (inside it and outside it) more thoroughly. Same can be said about all of these database functions to increase understandability :fireworks:

File: src/Server/Game.hs [hunk 8]
```diff
@@ -12,159 +12,166 @@ import Hastings.Utils
 import Hastings.ServerUtils
 import LobbyTypes
 
+import qualified Hastings.Database.Game as GameDB
+import qualified Hastings.Database.Player as PlayerDB
+import qualified Hastings.Database.Fields as Fields
+import qualified Database.Esqueleto as Esql
```
Comment (Kalior) [inline src/Server/Game.hs]: Separate the import of our Database from Esqueleto.

File: test/Hastings/ServerUtilsTests.hs [hunk 9]
```diff
@@ -1,115 +0,0 @@
--- |Tests functions from "Hastings.ServerUtils".
--- Does not test simple functions such as those running a 'find'
--- or simple 'map's.
-module Hastings.ServerUtilsTests where
```
Comment (Kalior) [inline test/Hastings/ServerUtilsTests.hs:4]: rip :skull:

**Comment by Kalior** [comment]
I added some comments to the code, add some tests, but otherwise it looks fine.

**Comment by Kalior** [comment]
Looks good, a bit bothered by the fact that most tests have to run multiple methods from the database module, but I suppose that is hard to get away from when dealing with a database.

**Comment by Kalior** [comment]
Why are the games created with some weird string as name?

**Comment by mattiasnilsen** [comment]
What should the default names for a game be then?

**Comment by Kalior** [comment]
Something that isn't mega strange, like GameName or default or something

**Comment by Kalior** [comment]
Perfect, will merge now :partly_sunny:

**Pull request merged**